### ISSUE TITLE
Extend Pattern Matcher to Match Equivalent Function Invocations

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#104391]
### 🐛 Describe the bug

If you add a print statement to `should_pad_addmm` in `pad_mm`.py :

```
--- a/torch/_inductor/fx_passes/pad_mm.py
+++ b/torch/_inductor/fx_passes/pad_mm.py
@@ -82,6 +82,7 @@ def addmm_pattern(input, mat1, mat2, beta, alpha):
 
 
 def should_pad_addmm(match):
+    print("TESTING FOR REPLACEMENT")
     mat1, mat2, input = fetch_fake_tensors(match, ("mat1", "mat2", "input"))
     return should_pad_common(mat1, mat2, input) and should_pad_bench(
         mat1, mat2, torch.ops.aten.addmm, input=input 
```

and then run the following script:

```
import torch

@torch.compile()
def test_mm_padding(inp, x, y):
    return torch.ops.aten.addmm(inp, x, y)

@torch.compile()
def test_mm_padding2(inp, x, y):
    return torch.ops.aten.addmm(inp, x, y, alpha=1.0, beta=1.0)

inps = [torch.rand([20]), torch.rand([20, 20]), torch.rand([20, 20])]
inps = [x.to(device="cuda") for x in inps]

print("Running first")
test_mm_padding(*inps)

print("Running second")
test_mm_padding2(*inps)
```

You'll see that only second addmm invocation which explicitly passes in alpha and beta get matched, even though both addmm invocations are equivalent. That is because we are matching on exact [# of args and kwargs](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/pattern_matcher.py#L343-L347). 

We should extend the pattern matcher so that equivalent invocations get matched. A relevant function here is [normalize_function](https://github.com/pytorch/pytorch/blob/732067e5c30519df5c8906913825bd905e9185fa/torch/fx/operator_schemas.py#L247). 

A couple things you may run into:

- there's an annoying thing where pytorch replaces "self" with "input
- this might break existing replacement patterns that are directly introspecting the fx node. replacement patterns should mostly be fine but graph patterns are more likely to break. we could have an option to the patterns that would canonicalize the calling convention to the convention that its been specified as. There aren't tooooo many existing patterns so it shouldn't be terrible to update them.

shout out @aakhundov for spotting this !

### Versions

master

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

**Comment by aakhundov** [comment]
@eellison while trying to understand the root cause of the issue, I've stumbled upon something potentially interesting.

As you'd noted, with `alpha=1.0, beta=1.0` the `pad_mm` pattern is matched. However, when I change the values to `alpha=1, beta=1`, the pattern is not matched anymore! The reason is that the default scalar values (as specified in the `native_functions.yaml`) are removed from the signature during normalization / canonicalization (and, indeed, for the `_addmm_activations` the default values of `alpha` and `beta` are specified as `1`, not `1.0` [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L6630,L6638)). Dropping of the kwargs with the default values seems to be happening within [this call](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/aot_autograd.py#L1371) to `make_fx` in the `aot_autograd.py`.

So one alternative might be extending `make_fx` to drop values which are equal to but of different type than the specified defaults of the kwargs. Although, what you've suggested---normalizing the function calls in the graph before matching, possibly when a special configuration flag is set---seems to give more flexibility in defining the patterns (e.g., one could then define the pattern with `alpha=alpha, beta=beta` as it's [done now](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/pad_mm.py#L81) in `pad_mm` without caring about default values). So, I guess, I'll proceed with the suggested `normalize_function` approach first.

**Comment by eellison** [comment]
sounds good ! Per Edward Yang, default values which are passed into operators are not passed along to `__torch_dispatch__` because of Forward-Compatibility reasons. You want adding a new default argument to not be bc breaking to torch dispatch implementers
